pytential/symbolic/stokes.py

```diff
@@ -55,28 +56,80 @@ class StokesletWrapper:
 
-        elif dim == 3:
-            self.kernel_dict = {
-                        (2, 0, 0): StokesletKernel(dim=3, icomp=0, jcomp=0),
-                        (1, 1, 0): StokesletKernel(dim=3, icomp=0, jcomp=1),
-                        (1, 0, 1): StokesletKernel(dim=3, icomp=0, jcomp=2),
-                        (0, 2, 0): StokesletKernel(dim=3, icomp=1, jcomp=1),
-                        (0, 1, 1): StokesletKernel(dim=3, icomp=1, jcomp=2),
-                        (0, 0, 2): StokesletKernel(dim=3, icomp=2, jcomp=2)
-                               }
+        self.kernel_dict = {}
 
-        else:
-            raise ValueError("unsupported dimension given to StokesletWrapper")
+        self.base_kernel = BiharmonicKernel(dim=dim)
+
+        if not use_biharmonic:
+            for i in range(dim):
+                for j in range(i, dim):
+                    self.kernel_dict[(i, j)] = StokesletKernel(dim=dim, icomp=i,
+                                                               jcomp=j)
+
+            for i in range(dim):
+                for j in range(i):
+                    self.kernel_dict[(i, j)] = self.kernel_dict[(j, i)]
+
+    def _get_stokeslet_int_g(self, icomp, jcomp, density, mu_sym, qbx_forced_limit,
+            deriv_dirs):
+
+        """
+        Returns the Integral of the Stokeslet kernel given by `icomp` and `jcomp`
+        and its derivatives.
+        Note that,
+
+            Stokeslet(icomp, jcomp)
+                = -1/mu_sym d/(d x_icomp) d/(d x_jcomp) BiharmonicKernel
+
+        for icomp != jcomp. Also note that for 3D,
+
+            Stokeslet(i, i)
+                = -1/mu_sym sum(-d^2/(dx_j^2 BiharmonicKernel for j != i).
+
+        For 2D,
+
+            Stokeslet(i, i)
+                = -1/mu_sym (sum(d^2/dx_j^2 BiharmonicKernel for j != i)
+                             -3/(8*pi)).
+        """
+
+        if not self.use_biharmonic:
+            knl = self.kernel_dict[(icomp, jcomp)]
+            for deriv_dir in deriv_dirs:
+                knl = AxisTargetDerivative(deriv_dir, knl)
+            return sym.IntG(knl, density,
+                    qbx_forced_limit=qbx_forced_limit, mu=mu_sym)
+
+        def func(knl, deriv_dirs):
+            for deriv_dir in deriv_dirs:
+                knl = AxisTargetDerivative(deriv_dir, knl)
+
+            res = sym.IntG(knl, density,
+                    qbx_forced_limit=qbx_forced_limit)
+            return res
+
+        mult = -1 / mu_sym
+
+        if icomp != jcomp:
+            return mult * func(self.base_kernel,
+                deriv_dirs=(deriv_dirs + [icomp, jcomp]))
+
+        const = 0
+        if self.dim == 2 and not deriv_dirs:
+            from math import pi
+            from pytential.symbolic.primitives import as_dofdesc, DEFAULT_SOURCE
+            const = -3/(8*pi) * sym.integral(self.dim, self.dim-1, density,
+                                             dofdesc=as_dofdesc(DEFAULT_SOURCE))
```

Comment:
@inducer, is this correct?

Reply:
Yes, it looks that way. One remaining issue is that the Stokeslet is being passed a density with the mean removed. This mean also involves an integral that needs to be told where to evaluate. I'm suspecting there might be a way to make the `ToTargetTagger` smarter to automatically tag things more appropriately.